airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py

```diff
@@ -0,0 +1,171 @@
+
+        logger.debug(f"try to open {file.uri}")
+        try:
+            result = smart_open.open(f"s3://{self.config.bucket}/{file.uri}", transport_params=params, mode=mode.value)
+        except OSError:
+            logger.warning(
+                f"We don't have access to {file.uri}. The file appears to have become unreachable during sync."
+                f"Check whether key {file.uri} exists in `{self.config.bucket}` bucket and/or has proper ACL permissions"
+            )
+        # see https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager for why we do this
+        try:
+            yield result
+        finally:
+            result.close()
+
+    @staticmethod
+    def _is_folder(file) -> bool:
+        return file["Key"].endswith("/")
+
+    def _page(
+        self, s3: BaseClient, globs: List[str], bucket: str, prefix: Optional[str], seen: Set[str], logger: logging.Logger
+    ) -> Iterable[RemoteFile]:
+        """
+        Page through lists of S3 objects.
+        """
+        total_n_keys_for_prefix = 0
+        kwargs = {"Bucket": bucket}
+        while True:
+            response = s3.list_objects_v2(Bucket=bucket, Prefix=prefix) if prefix else s3.list_objects_v2(Bucket=bucket)
+            key_count = response.get("KeyCount")
+            total_n_keys_for_prefix += key_count
+            logger.info(f"Received {key_count} objects from S3 for prefix '{prefix}'.")
+
+            if "Contents" in response:
+                for file in response["Contents"]:
+                    if self._is_folder(file):
+                        continue
+                    remote_file = RemoteFile(uri=file["Key"], last_modified=file["LastModified"])
+                    if self.file_matches_globs(remote_file, globs) and remote_file.uri not in seen:
+                        seen.add(remote_file.uri)
+                        yield remote_file
+            else:
+                logger.warning(f"Invalid response from S3; missing 'Contents' key. kwargs={kwargs}.")
+
+            if next_token := response.get("NextContinuationToken"):
+                kwargs["ContinuationToken"] = next_token
+            else:
+                logger.info(f"Finished listing objects from S3 for prefix={prefix}. Found {total_n_keys_for_prefix} objects.")
+                break
+
+
+# Used to specify anonymous (unsigned) request signature
+class UNSIGNED:
+    def __copy__(self):
+        return self
+
+    def __deepcopy__(self, memodict):
+        return self
+
+
+def _get_s3_compatible_client_args(config: Config) -> dict:
+    """
+    Returns map of args used for creating s3 boto3 client.
+    """
+    client_config = ClientConfig(signature_version=UNSIGNED())
+    client_kv_args = {"config": client_config}
+    client_kv_args.update(
+        {
+            "endpoint_url": config.endpoint,
+            "use_ssl": True,
+            "verify": True,
+            "config": ClientConfig(s3={"addressing_style": "auto"}),
```

Comment:
I'm probably missing something: why do we set `client_kv_args = {"config": client_config}` if we update `client_kv_args["config"]` right after with `ClientConfig(s3={"addressing_style": "auto"})`?

Reply:
You're right, this is awkward (a remnant of copy/pasting & an incomplete refactor). Cleaned it up.